setup.cfg

```diff
@@ -33,6 +33,7 @@ install_requires =
     astropy-iers-data>=0.2023.8.28.0.30.15
     PyYAML>=3.13
     packaging>=19.0
+    typing_extensions>=4.6.2
```

Comment:
This is a backport package for new functionality being added in Python 3.12. So I don't think it's unreasonable to added to get that functionality now. If we go this way a guard to not install this package on 3.12+ should be added.

Alternatively, we could adopt these changes for Python 3.12+ only and maintain the current system for now. The downside to this is that we would have to maintain the current system in addition to the proposed changes here.

Reply:
Can we defer this PR until numpy 2.0 is stable and Python 3.12 is out? Then my mood would be better and I might view this more favorably. 😜 

Right now this functionality is pretty stable so I don't see any rush in replacing it.

Reply:
Totally fine to defer. I agree there's no rush. It's just better to think about this way before python 3.12 is our minimum supported version. And since Python itself develops `typing_extensions` (look at where the files are hosted on GitHub), this package affords us the opportunity to build for the future well before it arrives.

Reply:
Yes, if it is going to be included in Python 3.12, then as William said, we need a qualifier to not install the package for that Python onwards. And it would be quite a while before our minversion of Python is 3.12, so indeed we have plenty of time.

Reply:
I have a similar (in spirit) suggestion that avoids to python version checks at runtime.

1. We use `typing_extensions` for all python versions since `typing_extensions` has the python version check built in (as in `typing_extensions` does
```python
if PYTHON_VERSION >= 3.12:
    from typing import deprecated
else:
    def deprecated(...
```
).

2. [Open an issue on ruff](https://github.com/astral-sh/ruff/issues/7285) asking for a pyupgrade-style code that auto-fixes the import when appropriate, aka when we change ruff's target version to 3.12 because our minimum supported python is 3.12+.

Reply:
Your suggested rule for ruff sounds like a good idea to me (although it might be cool to submit it to `pyupgrade` too), but one thing that ruff will probably never do is clean up unused dependencies from the package metadata. So, as stable as `typing_extensions` might be I feel like it's still preferable to avoid installing it alongside versions of Python that don't need it, so in the future one can *see* that the dependency isn't needed anymore just by reading metadata (`pyproject.toml` in the near future). In that scenario your proposed pyupgrade rule would still help staying on top of the game, but we wouldn't have to completely rely on it.

Reply:
> unused dependencies

We're only going to be adding more typing in the future, not taking it away. I would be very surprised if `typing_extensions` is ever unused.

Reply:
Yeah, realistically I think you're right. But if we limit our use to stuff that has landed in the standard lib (as I think we should), then the extension will not be needed for the latest version of Python at any point in time. I have a preference to make that explicit, but your argument is sound.

Reply:
>  I anticipate that we will want typing_extensions installed always to backport those newer features.

I still don't understand why we cannot tell setup to not install it for Python >= 3.12. I must be missing an important point amid these discussions. Having it explicit makes removing dead code easier when our minversion of Python is finally 3.12.

Reply:
@pllim to back up @nstarman , I can say that in my experience, `typing_extensions` tends to stick as a dependency because each year brings a bunch of typing features that are only accessible through it for most Python versions, except the most recent one. Because of this, I would not anticipate it to be dropped when Python 3.12 becomes the new minimum. That said, I'm still in favour of making version specific requirements as explicit as possible, and avoid making an exception for `typing_extensions` solely based on current expectations.

Reply:
Done. This will have to introduce python version info checks in the import sections.

Given that `typing_extensions` is an official backport package, really small at ~73 kB in python (and 33 kB wheels), and probably sticking around forever, I was kind of hoping to avoid piling up version checks for determining whether `typing_extensions` is installed. We might end up with > 33 kB in those 😢.